### DISCUSSION TITLE
Await async remote procedure calls

### DISCUSSION THREAD

**Discussion by vojtechmusilek** [FirstGearGames/FishNet#202] [Feature Request]
Currently after calling Rpc code continues before Rpc finishes.
Add option to await this call when Rpc is async.
Example:
```csharp
async void Something()
{
  ...
  await ServerSomething()
  ...
}

[ServerRpc]
async void ServerSomething()
{
  ...
}
```

**Comment by FirstGearGames** [comment]
I believe this is on the roadmap under wishlist. After sitting on this
awhile I probably won't be implementing it to be honest. There's a lot of
technical reasons why this doesn't work.

**Reply by vojtechmusilek** [reply]
My workaround was basically looping around back to clients and setting some bool to true causing them to continue.
But I understand it would be probably hard to implement and would require some big changes.

Maybe this system could be somehow abstracted but I didn't give it much thought yet.

Workaround example:
```csharp
bool rpcFinished;

void Something()
{
  rpcFinished = false
  ServerSomething()
  
  ... //wait until rpcFinished is true. this can be done with some Coroutine or using lib UniTask.WaitUntil
  
  ... //logic after rpc finishes
}

[ServerRpc]
async void ServerSomething()
{
  ... //logic on server
  
  ClientSomething()
}

[ObserversRpc]
async void ClientSomething()
{
  ... //logic on all clients
  
  rpcFinished = true
}

```

**Reply by FirstGearGames** [reply]
I can absolutely understand the convenience but there's a considerable amount of work required for this change.
